### ISSUE TITLE
IndexOutOfRangeException exception in Matrix class, .First method when using lambda

### ISSUE THREAD

**Issue by bebenins** [accord-net/framework#696]
Hello,
I think this should be gracefully handled in .First method.
Here's what I have...

I need to find an index of a specific Point.X in an array of Points.  So I use this expression: 
```
    double myX = 700;
    int index = fft.First(p => p.X >= myX);
```
It works just fine if myX isn't grater than the maximum Point.X in fft array.  But if my fft array Point.X is less than myX, I get an exception:

> System.IndexOutOfRangeException occurred
  HResult=-2146233080
  Message=Index was outside the bounds of the array.
  Source=Accord.Math
  StackTrace:
       at Accord.Math.Matrix.First[T](T[] data, Func`2 func)
  InnerException: 

What is the easiest/elegant "patch" I can do for now?
Thank you!

Edit:  Just in case, this is using Accord.Math 3.4.0.0

**Comment by cesarsouza** [comment]
Hi @bebenins,

In this case, the exception happens because the point could not be found right?

I would like to avoid returning -1 in this case because -1 can be interpreted as a correct index if you use it with the .Get() and .Set() functions (where it would mean "return the last element in the array").

I guess it would be better to add a FirstOrDefault method that could handle this case, where the output could be an index, or null. But it might take a while until it could be added into a release (or even pre-release version, since the last one was published yesterday).

If you want a quick (but not elegant) fix for now, you can add an extension method like this to your project:

```
public static int? FirstOrDefault<T>(this T[] array, Func<T, bool> pred)
{
    try 
    {
        return array.First(array, pred);
    }
    catch (IndexOutOfRangeException ex)
    {
        return null;
    }
}
```

Then call it instead of .First().

PS: I have just typed the code above here in GitHub and didn't attempt to compile, so it might contain errors.

Regards,
Cesar

**Comment by bebenins** [comment]
Thanks Cesar!
Yes I'm fairly certain the reason for exception is because value cannot be found.  I agree that -1 isn't ideal and FirstOrDefault support would be awesome.
Thanks for the suggestion in the mean time!

**Comment by cesarsouza** [comment]
Alternatively, you can also call ```int[] indices = fft.First(p => p.X >= myX, firstOnly: true)```, then check if the length of the indices vector is zero.

**Comment by cesarsouza** [comment]
Fixed in 3.7.0. You can now use the .FirstOrNull() method to find the first value in the vector, or null, in case it is not present. Alternatively, you can also use .Find() and check whether the length of the resulting vector is zero before attempting to retrieve its first element.
